### ISSUE TITLE
Run the game on macOS

### ISSUE THREAD

**Issue by Lnd-stoL** [Swartz27/xray-16#1104]
Although the engine can be built successfully on macOS, the game won't run.

**Comment by Lnd-stoL** [comment]
- #1105 
- #1139

**Comment by TmLev** [comment]
Should be closed now?..

**Comment by Xottab-DUTY** [comment]
@Lnd-stoL, what do you think? Are we enough or there's few minor/major fixes to do?

**Comment by Xottab-DUTY** [comment]
~~Well, we actually launched the game on macOS, so, let's close this as done!~~
Lets's stabilize...

**Comment by banq007** [comment]
Would there be any macos binaries as artifacts in github actions?

**Comment by tsunamaru** [comment]
I managed to build and run engine on MacOS 13 (macbook pro 16 2021, m1 max, 32gb) and play COP, but game starts being very laggy when there is any npcs around (drops from 120+ fps to 10-15). Also lights is broken in many parts (flickering, incorrect brighness on objects, etc)

UPD: building engine with `-DLUAJIT_DISABLE_JIT=ON` resolves fps drop issue.

**Comment by TmLev** [comment]
@tsunamaru can you share the details on how you managed to build it?

**Comment by tsunamaru** [comment]
@TmLev sure. I mostly followed steps from https://github.com/OpenXRay/xray-16/blob/dev/.github/workflows/cibuild.yml

1. Clone repository with submodules, `cd` inside
2. `git submodule deinit Externals/cryptopp`
3. `brew install sdl2 glew lzo jpeg openal-soft libogg libvorbis theora`
4. `CFLAGS="-w" CXXFLAGS="-w" cmake -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_UNITY_BUILD=ON -DLUAJIT_DISABLE_JIT=ON -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15`
5. `cmake --build build --config Release --parallel 8`
6. copy gamedata to `~/Library/Application Support/GSC Game World/S.T.A.L.K.E.R. - Call of Pripyat`
7. now you can start `./bin/x64/Release/xr_3da`

**Comment by Xottab-DUTY** [comment]
> @tsunamaru can you share the details on how you managed to build it?

I remember you had issues with building process... @johncurley have had fixed some issues with it, so you might want to try it.

**Comment by TmLev** [comment]
> @TmLev sure. I mostly followed steps from https://github.com/OpenXRay/xray-16/blob/dev/.github/workflows/cibuild.yml

Thanks!

> > @tsunamaru can you share the details on how you managed to build it?
> 
> I remember you had issues with building process... @johncurley have had fixed some issues with it, so you might want to try it.

Sure, will try again.

**Comment by TmLev** [comment]
Yeah, no luck again. 

I'm running macOS 12.4 on M1 Pro and compiler can't seem to find `stdio.h`:

```logs
In file included from /Users/lev/repos/xray-16/Externals/LuaJIT/src/ljamalg.c:19:
/Users/lev/repos/xray-16/Externals/LuaJIT/src/lauxlib.h:13:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
```

Full logs: https://gist.github.com/TmLev/746c0bda98d2ed01c0cb6ab6733bfc87

I've also tried `export MACOSX_DEPLOYMENT_TARGET=12.4` (with respecting `-DCMAKE_OSX_DEPLOYMENT_TARGET=12.4`) and [`export SDKROOT=$(xcrun --sdk macosx --show-sdk-path)`](https://stackoverflow.com/questions/51761599/cannot-find-stdio-h)

Will spend some time trying to fix it.

**Comment by tsunamaru** [comment]
@TmLev most likely you have some issues with xcode installed, go grab latest version from apple site; also make sure xcode commandline tools is installed and updated as well. Just fyi mine xcode is 14.3 and macos 13.4

**Comment by TmLev** [comment]
> @TmLev most likely you have some issues with xcode installed, go grab latest version from apple site; also make sure xcode commandline tools is installed and updated as well. Just fyi mine xcode is 14.3 and macos 13.4

You're totally right, for some reason it wasn't even installed (even though I did that a hundred times at this point).

Holy shit, it actually runs with decent performance!!!

**Comment by johncurley** [comment]
There's still some issues with the OpenGL renderer on MacOS around Indexed Drawing. I'd eventually like to get around to it but for now a lot of refactoring is going on and I also have no idea how to fix the MacOS problem. So I'll have to devote some time to look into it but am currently working on the sound system and editors.

**Comment by tsunamaru** [comment]
There is a few issues I observed during the play on macos (besides from broken rendering):  
1. Rain ignores any walls, making it everywhere inside the buildings.
2. Enemy ai also ignores any walls on a way to player, it can see and shoot you through everything. Makes you very frustrating when you encounter controller from which you can't hide 😅
3. Shooting npc from close distance causes stutters and fps drops.

**Comment by Xottab-DUTY** [comment]
> Would there be any macos binaries as artifacts in github actions?

@banq007, I've created a todo list for improvements in Actions, macOS binaries can be watched there too: #1403

**Comment by Xottab-DUTY** [comment]
> Rain ignores any walls, making it everywhere inside the buildings.

That's original game bug that doesn't seem to be trivial to fix :(

**Comment by yohjimane** [comment]
> Rain ignores any walls, making it everywhere inside the buildings

@Xottab-DUTY I fixed this bug for r3 in the Shader Support branch, with this commit. 
See: 5c29f2e94196a083d8e4a63ef4e5a9cc0fa7b5e6

Should be able to apply this fix to the rest of the shaders too

**Comment by tsunamaru** [comment]
@Xottab-DUTY 
> That's original game bug that doesn't seem to be trivial to fix :(

Mhm, to be honest I don't remember this behavior in original COP (1.6.02), at least not in such form as I've seen during my play on MacOS, where not only rain is visible inside buildings, but all surfaces become wet as well.

**Comment by Xottab-DUTY** [comment]
There might be issues common with other platforms (e.g. binaries compiled with Clang are less stable).
> Rain ignores any walls, making it everywhere inside the buildings.

This is also a common issue...

So, I'll close this since the game runs.
